src/controller/CHIPDeviceControllerFactory.cpp

```diff
@@ -0,0 +1,284 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implementation of CHIP Device Controller Factory, a utility/manager class
+ *      that vends Controller objects
+ */
+
+#include <controller/CHIPDeviceControllerFactory.h>
+
+#include <lib/support/ErrorStr.h>
+
+#if CONFIG_DEVICE_LAYER
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConfigurationManager.h>
+#endif
+
+using namespace chip::Inet;
+using namespace chip::System;
+using namespace chip::Credentials;
+
+namespace chip {
+namespace Controller {
+
+CHIP_ERROR DeviceControllerFactory::Init(FactoryInitParams params)
+{
+
+    // SystemState is only set the first time init is called, after that it is managed
+    // internally. If SystemState is set then init has already completed.
+    if (mSystemState != nullptr)
+    {
+        ChipLogError(Controller, "Device Controller Factory already initialized...");
+        return CHIP_NO_ERROR;
+    }
+
+    mListenPort      = params.listenPort;
+    mStorageDelegate = params.storageDelegate;
+
+    CHIP_ERROR err = InitSystemState(params);
+
+    return err;
+}
+
+CHIP_ERROR DeviceControllerFactory::InitSystemState()
+{
+    FactoryInitParams params;
+    if (mSystemState != nullptr)
+    {
+        params.systemLayer = mSystemState->SystemLayer();
+        params.inetLayer   = mSystemState->InetLayer();
+#if CONFIG_NETWORK_LAYER_BLE
+        params.bleLayer = mSystemState->BleLayer();
+#endif
+    }
+
+    return InitSystemState(params);
+}
+
+CHIP_ERROR DeviceControllerFactory::InitSystemState(FactoryInitParams params)
+{
+    if (mSystemState != nullptr && mSystemState->IsInitialized())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (mSystemState != nullptr)
+    {
+        mSystemState->Release();
+        chip::Platform::Delete(mSystemState);
+        mSystemState = nullptr;
+    }
+
+    DeviceControllerSystemStateParams stateParams;
+#if CONFIG_DEVICE_LAYER
+    ReturnErrorOnFailure(DeviceLayer::PlatformMgr().InitChipStack());
+
+    stateParams.systemLayer = &DeviceLayer::SystemLayer();
+    stateParams.inetLayer   = &DeviceLayer::InetLayer;
+#else
+    stateParams.systemLayer = params.systemLayer;
+    stateParams.inetLayer   = params.inetLayer;
+    ChipLogError(Controller, "Warning: Device Controller Factory should be with a CHIP Device Layer...");
```

Comment:
What happens if this is printed?

Reply:
The warning is just noting that the Factory was initialized without a Device Layer. The CHIP stack cannot be restarted once the last controller is released.
Android doesn't currently have a Device Layer, it manages these system and inet Layers on its own. Which means restarts of the CHIP stack are not going work as the ownership lies with the Android wrapper. This is true even prior to this PR. 

related: I left a TODO in `FactoryInitParams` that also talks about how we need to remove the platform layers from the init params.
```
// TODO everything other than the storage delegate here should be removed.
// We're blocked because of the need to support !CHIP_DEVICE_LAYER
```